anchor-service/src/anchor_batch.rs

```diff
@@ -14,31 +21,121 @@ use crate::{
+    anchor_interval: Duration,
+    anchor_batch_size: u64,
 }
 
 impl AnchorService {
     /// Create a new AnchorService.
     pub fn new(
         tx_manager: Arc<dyn TransactionManager>,
-        _event_service: Arc<dyn Store>,
-        _batch_linger_time: Duration,
+        event_service: Arc<dyn Store>,
+        pool: SqlitePool,
+        node_id: NodeId,
+        anchor_interval: Duration,
+        anchor_batch_size: u64,
     ) -> Self {
         Self {
             tx_manager,
-            _event_service,
-            _batch_linger_time,
+            event_service,
+            high_water_mark_store: HighWaterMarkStore::new(pool),
+            node_id,
+            anchor_interval,
+            anchor_batch_size,
+        }
+    }
+
+    /// Run the AnchorService:
+    /// - Get anchor requests from the AnchorClient
+    /// - Deduplicate the anchor requests
+    /// - Anchor the batch using a Transaction Manager
+    /// - Store the TimeEvents using the AnchorClient
+    ///
+    /// This function will run indefinitely, or until the process is shutdown.
+    pub async fn run(&mut self, shutdown_signal: impl Future<Output = ()>) {
+        let mut interval = interval(self.anchor_interval);
+        let shutdown_signal = shutdown_signal.fuse();
+        pin_mut!(shutdown_signal);
+
+        info!("anchor service started");
+
+        loop {
+            let tick_future = interval.tick().fuse();
+            pin_mut!(tick_future);
+
+            match select(tick_future, &mut shutdown_signal).await {
+                Either::Left((_, _)) => {
+                    // Interval tick occurred
+                    if let Err(e) = self.process_next_batch().await {
+                        error!("Error processing batch: {:?}", e);
+                    }
+                }
+                Either::Right((_, _)) => {
+                    break;
+                }
+            }
+        }
+        info!("anchor service stopped");
+    }
+
+    async fn process_next_batch(&mut self) -> Result<()> {
+        // Pass the anchor requests through a deduplication step to avoid anchoring multiple Data Events from the same
+        // Stream.
+        let high_water_mark = self
+            .high_water_mark_store
+            .high_water_mark()
+            .await
+            .expect("error getting high water mark from database");
+        // Get the next batch of anchor requests
+        let anchor_requests: Vec<AnchorRequest> = match self
+            .event_service
+            .events_since_high_water_mark(
+                self.node_id,
+                high_water_mark,
+                self.anchor_batch_size as i64,
+            )
+            .await
+        {
+            Ok(requests) => IndexMap::<Cid, AnchorRequest>::from_iter(
+                requests.into_iter().map(|request| (request.id, request)),
+            )
+            .into_values()
+            .collect(),
+            Err(e) => {
+                return Err(e);
```

Comment:
if this fails, what happens? it seems like it will just keep failing forever or am I misunderstanding?